src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -161,6 +166,95 @@ bool Compiler::optRedundantBranch(BasicBlock* const block)
+                                            // even though the IR does not support this.
+                                            //
+                                            for (auto vnRelation : vnRelations)
+                                            {
+                                                const ValueNum relatedVN = vnStore->GetRelatedRelop(pVN, vnRelation);
+
+                                                if ((relatedVN != ValueNumStore::NoVN) && (relatedVN == treeNormVN))
+                                                {
+                                                    vnRelationMatch = vnRelation;
+                                                    matched         = true;
+
+                                                    // If dom predicate is wrapped in EQ(*,0) then a true dom
+                                                    // predicate implies a false branch outcome, and vice versa.
+                                                    //
+                                                    // And if the dom predicate is GT_NOT we reverse yet again.
+                                                    //
+                                                    reverseSense = (oper == GT_EQ) ^ (predOper == GT_NOT);
+
+                                                    // We only get partial knowledge in these cases.
+                                                    //
+                                                    //   AND(p1,p2) = true  ==> both p1 and p2 must be true
+                                                    //   AND(p1,p2) = false ==> don't know p1 or p2
+                                                    //    OR(p1,p2) = true  ==> don't know p1 or p2
+                                                    //    OR(p1,p2) = false ==> both p1 and p2 must be false
+                                                    //
+                                                    if (predOper != GT_NOT)
+                                                    {
+                                                        canInferFromFalse = reverseSense ^ (predOper == GT_OR);
+                                                        canInferFromTrue  = reverseSense ^ (predOper == GT_AND);
+                                                    }
+
+                                                    JITDUMP("Inferring predicate value from %s\n",
+                                                            GenTree::OpName(predOper));
+                                                    break;
+                                                }
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
```

Comment:
> I am tempted to recast this as a general VN utility where we can ask if knowing the value of a VN implies knowing the value of another VN; there are lots of other inferences we can draw of this kind. But I'm not sure how to structure this without it getting overly complicated.

I think this would be great, and in any case it would be nice to extract this to a function to avoid some of the nesting.

Reply:
> > I am tempted to recast this as a general VN utility where we can ask if knowing the value of a VN implies knowing the value of another VN; there are lots of other inferences we can draw of this kind. But I'm not sure how to structure this without it getting overly complicated.
> 
> I think this would be great, and in any case it would be nice to extract this to a function to avoid some of the nesting

Let me get the logic right and then I'll look into refactoring.